app/rdf/query-hierarchies.ts

```diff
@@ -73,24 +73,39 @@ const toTree = (
   return sortChildren(results.map((r) => serializeNode(r, 0)).filter(truthy));
 };
 
-const findHierarchiesForDimension = (
+const getDimensionHierarchies = async (
   cube: Cube,
   dimensionIri: string,
   locale: string
 ) => {
-  const newHierarchies = uniqBy(
-    cube.ptr
-      .any()
-      .has(ns.sh.path, rdf.namedNode(dimensionIri))
-      .has(ns.cubeMeta.inHierarchy)
-      .out(ns.cubeMeta.inHierarchy)
-      .toArray(),
-    (x) => getName(x, locale)
-  );
-  if (newHierarchies) {
-    return newHierarchies;
+  const cubeIri = cube.term?.value;
+
+  if (!cubeIri) {
+    throw new Error("Cube must have an iri!");
   }
-  return [];
+
+  // Can't rely on the cube here, because sometimes it might contain duplicate
+  // hierarchies, due to fetching both cube and shape when cube is initialized.
+  // Here we are only interested in shape hierarchies, that's why we don't have
+  // to initialize the cube, but rather just fetch its shape.
+  const hierarchyCube = new Cube({
+    source: cube.source,
+    term: rdf.namedNode(cubeIri),
+  });
+  // Need to set the shape query manually, because we didn't initialize the cube,
+  // thus can't access its observationConstraint.
+  hierarchyCube.shapeQuery = () => {
+    return `DESCRIBE <${cube.out(ns.cube.observationConstraint)}>`;
+  };
+  await hierarchyCube.fetchShape();
```

Comment:
two questions:
- does this trigger yet another query?
  - i.e. might this have an impact on performance?
  - can separate queries be avoided? i.e. by filtering cubeIRI to only the single relevant cube?
  - nb: is this query traced in debug mode? (once the let let's you reuse your client)

Reply:
I refactored the PR a bit and modified the `Cube` class from cube-view-query library, so we can differentiate if given triple comes from describing the `cube` or its `shape`. Now we won't fire any additional query :)

Reply:
with your refactored changes there is still the shape for the expired cube being fetched somehow... 
```sparql
DESCRIBE <https://environment.ld.admin.ch/foen/nfi/nfi_C-90/cube/2023-2/shape/>
```
do you init that cube somewhere before updating to non-expired?

Reply:
I was testing with [this chart](https://visualize.admin.ch/de/v/IYdcAI0tH0Io?dataSource=Prod), which is based on v1 of the cube, while the newest one is v3 – and with the change we fetch the correct shape.

<img width="478" alt="Screenshot 2023-11-10 at 13 37 53" src="https://github.com/visualize-admin/visualization-tool/assets/52032047/c7bba5a8-cd21-40fb-a0a9-68b8c2d2e08d">

But it's true that we might fetch some data related to cube that won't be used (as a newer cube is available) – we need it to determine if we need to fetch newer cube, so I'm not sure we can avoid this. We might not use full `DESCRIBE` query, but rather smaller one to just fetch things we need, but that would be a performance improvement not related to hierarchy bug 👍 

Could you share more information on how you encountered below?

```
DESCRIBE <https://environment.ld.admin.ch/foen/nfi/nfi_C-90/cube/2023-2/shape/>
```

Reply:
@bprusinowski you can get such information as well using
- mitmproxy
- introduce full traceability (once the lib is ready)
sure let's tackle this in the perf context. :+1: 
NB: did you already start with the perf improvement for preview? https://gitlab.ldbar.ch/bafu/visualize/-/issues/547

Reply:
Ah, I missed the point, I thought that we fetch wrong shape when querying dimensions, and it was about generally not fetching outdated shapes 🤦‍♂️ 

I tackled this here: https://github.com/visualize-admin/visualization-tool/pull/1260/commits/d08bd37e770b62f5c3e7e8bcf5a04926f23884fc

Reply:
I've seen that Tomasz updated the packages, but didn't yet get around to implementing them 😄

Reply:
@bprusinowski as a follow up to this PR it would be helpful to 
1. introduce tracability based on newly released version
2. refactor preview completely based on proposed optimization from GL issue (1 single CONSTRUCT query) and fine tune that query :smile: